### DISCUSSION TITLE
STMP problem on v2.1 install regardless of adapter

### DISCUSSION THREAD

**Discussion by anttikivi** [plausible/analytics#4149] [Self-Hosted Support]
Hi!

I’m installing a Plausible CE v2.1 on a new server. However, I don’t seem to get the mailer working regardless of the SMTP adapter I use. Without the new adapter shipped with v2.1 (without `MAILER_ADAPTER` in `plausible-conf.env`), test sending email from Plausible with the following commands:

```elixir
iex(plausible@fbe9c3dd799c)1> Application.get_env(:plausible, Plausible.Mailer)
[
  adapter: Bamboo.SMTPAdapter,
  server: "email-smtp.eu-north-1.amazonaws.com",
  hostname: "example.com",
  port: "587",
  username: "username",
  password: "password",
  tls: :if_available,
  allowed_tls_versions: [:tlsv1, :"tlsv1.1", :"tlsv1.2"],
  ssl: "false",
  retries: "2",
  no_mx_lookups: true
]
iex> email = Bamboo.Email.new_email(from: "admin@example.com", to: "example@example.com", subject: "test", text_body: "test")
%Bamboo.Email{
  from: "admin@example.com",
  to: "example@example.com",
  cc: nil,
  bcc: nil,
  subject: "test",
  html_body: nil,
  text_body: "test",
  headers: %{},
  attachments: [],
  assigns: %{},
  private: %{},
  blocked: false
}
iex> Plausible.Mailer.deliver_now(email)
```

I get the following error when running the code above (with real domains and email addresses of course):

```
** (CaseClauseError) no case clause matching: <<0>>
    (gen_smtp 1.2.0) /app/deps/gen_smtp/src/gen_smtp_client.erl:841: :gen_smtp_client.read_possible_multiline_reply/1
    (gen_smtp 1.2.0) /app/deps/gen_smtp/src/gen_smtp_client.erl:631: :gen_smtp_client.do_AUTH_each/5
    (gen_smtp 1.2.0) /app/deps/gen_smtp/src/gen_smtp_client.erl:556: :gen_smtp_client.try_AUTH/3
    (gen_smtp 1.2.0) /app/deps/gen_smtp/src/gen_smtp_client.erl:384: :gen_smtp_client.open_smtp_session/2
    (gen_smtp 1.2.0) /app/deps/gen_smtp/src/gen_smtp_client.erl:316: :gen_smtp_client.try_smtp_sessions/3
    (gen_smtp 1.2.0) /app/deps/gen_smtp/src/gen_smtp_client.erl:289: :gen_smtp_client.send_it/2
    (bamboo_smtp 4.2.2) lib/bamboo/adapters/smtp_adapter.ex:87: Bamboo.SMTPAdapter.deliver/2
    iex:3: (file)
```

The same configuration (with AWS SES on port 587 and SSL disabled) worked fine on Plausible v2.0.

When I’m using the new SMTP adapter (just added `MAILER_ADAPTER=Bamboo.Mua` in `plausible-conf.env`), testing the email with same code:

```elixir
iex(plausible@4950fbaad252)1> Application.get_env(:plausible, Plausible.Mailer)
[
  adapter: Bamboo.Mua,
  auth: [
    username: "username",
    password: "password"
  ],
  relay: "email-smtp.eu-north-1.amazonaws.com",
  port: 587
]
iex(plausible@4950fbaad252)2> email = Bamboo.Email.new_email(from: "admin@example.com", to: "example@example.com", subject: "test", text_body: "test")
%Bamboo.Email{
  from: "admin@example.com",
  to: "example@example.com",
  cc: nil,
  bcc: nil,
  subject: "test",
  html_body: nil,
  text_body: "test",
  headers: %{},
  attachments: [],
  assigns: %{},
  private: %{},
  blocked: false
}
iex(plausible@4950fbaad252)3> Plausible.Mailer.deliver_now(email)
```

I get the following error:

```
{:error,
 %Mua.TransportError{
   reason: {:tls_alert,
    {:unexpected_message,
     ~c"TLS client: In state hello_retry_middlebox_assert at ssl_gen_statem.erl:803 generated CLIENT ALERT: Fatal - Unexpected Message\n {unexpected_msg,\n     {internal,\n         {server_hello,\n             {3,3},\n             <<156,201,202,55,75,224,151,152,26,37,50,113,...>>,\n             <<191,136,249,229,37,77,160,34,70,157,19,...>>,\n             <<19,2>>,\n             0,\n             \#{server_hello_selected_version =>\n                   {server_hello_selected_version,{3,4}},\n               key_share =>\n                   {key_share_server_hello,\n                       {key_share_entry,secp256r1,<<4,51,...>>}},\n               pre_shared_key => undefined}}}}"}}
 }}
```

I used Caddy as reverse proxy but confirmed that these errors are also present when using Nginx with the Plausible-provided configuration with only the automatic Certbot modifications.

**Comment by ruslandoga** [comment]
👋 @anttikivi 

This discussion might be helpful: https://github.com/plausible/analytics/discussions/4084#discussioncomment-9363412

**tl;dr** you can try disabling middlebox_comp_mode and it should work with Mua:

1. Create an extra config file:

<sub><kbd>my_config.exs</kbd></sub>
```elixir
import Config

config :plausible, Plausible.Mailer,
  # in v2.1.1 and after this option is called :ssl, see my comment below (https://github.com/plausible/analytics/discussions/4149#discussioncomment-9725329)
  transport_opts: [middlebox_comp_mode: false]
```

2. Mount it into `plausible` container

<sub><kbd>[docker-compose.yml](https://github.com/plausible/community-edition/blob/v2.1.0/docker-compose.yml)</kbd></sub>
```diff
plausible:
  image: ghcr.io/plausible/community-edition:v2.1.0
+ volumes:
+   - ./my_config.exs:/app/my_config.exs
```

3. Make Plausible read it on startup

<sub><kbd>[plausible-conf.env](https://github.com/plausible/community-edition/blob/v2.1.0/plausible-conf.env)</kbd></sub>
```diff
+ EXTRA_CONFIG_PATH=/app/my_config.exs
```

**Reply by anttikivi** [reply]
Thank you! I tried the modifications but I still can’t get email sent. Now I get the following error:

```
** (exit) :badarg
    (kernel 9.2) gen_tcp.erl:229: :gen_tcp.connect/4
    (mua 0.1.5) lib/mua.ex:202: Mua.connect/5
    (mua 0.1.5) lib/mua.ex:114: Mua.easy_send_one/6
    (mua 0.1.5) lib/mua.ex:83: Mua.easy_send_any/6
    (bamboo_mua 0.1.4) lib/bamboo_mua.ex:56: Bamboo.Mua.deliver/2
    (bamboo 2.3.0) lib/bamboo/mailer.ex:208: Bamboo.Mailer.deliver_now/4
    iex:3: (file)
```

**Reply by ruslandoga** [reply]
Right. I'll fix it in Mua: https://github.com/ruslandoga/mua/issues/43

In the meantime, you can try using port 465 with SSL from the start:

<sub><kbd>my_config.exs</kbd></sub>
```elixir
import Config

config :plausible, Plausible.Mailer,
  protocol: :ssl,
  # in v2.1.1+ this option is named ssl, see my comment below (https://github.com/plausible/analytics/discussions/4149#discussioncomment-9725329)
  transport_opts: [middlebox_comp_mode: false]
```

<sub><kbd>[plausible-conf.env](https://github.com/plausible/community-edition/blob/v2.1.0/plausible-conf.env)</kbd></sub>
```diff
- SMTP_HOST_PORT=587
+ SMTP_HOST_PORT=465
```

**Reply by wheeleran** [reply]
Amazon SES will only accept 587, as far as i understand. this is rather frustrating.

**Reply by ruslandoga** [reply]
Starting in v2.1.1 `middlebox_comp_mode` and other `:ssl` options work with STARTTLS as well.

**Reply by ruslandoga** [reply]
<sub><kbd>my_config.exs</kbd></sub>
```elixir
import Config

config :plausible, Plausible.Mailer,
  ssl: [middlebox_comp_mode: false]
```

<sub><kbd>plausible-conf.env</kbd></sub>
```env
MAILER_ADAPTER=Bamboo.Mua
SMTP_HOST_PORT=587
```

**Reply by wheeleran** [reply]
@ruslandoga this is going to be difficult on my Synology NAS; there are all sort of issues currently with its docker Terminal, and finding a workaround is beyond my technical knowledge. it's such a pity, because email used to just WORK, and for whatever reason - for a few months now - fixing it is beyond me and i suppose others. is this the way it's going to be moving forward? or is it something that maybe a future update could possibly fix?

**Reply by Striffly** [reply]
I agree with this, Bamboo.MUA causes many problems for many users, myself included. Could it be possible to make Bamboo.SMTPAdapter functional again?

**Reply by ruslandoga** [reply]
> Could it be possible to make Bamboo.SMTPAdapter functional again?

It would be the same or worse :) And it's very unlikely to happen since it doesn't seem to be maintained anymore: https://github.com/fewlinesco/bamboo_smtp

To be clear, the new adapter hasn't really had any problems of its own yet. All of the issues so far have come from Erlang's TLS being too strict and failing to connect to misconfigured servers.

**Reply by wreidlinger** [reply]
@ruslandoga thanks for the steps to fix this issue.
I have made exactly the same steps but ran into an error. Maybe I did something wrong...
The used version is v2.1.1

My config:

- `docker-compose.yml` (plausible part of it)

```bash
# **************************************
# PLAUSIBLE SERVICE
# **************************************
  plausible:
    image: ghcr.io/plausible/community-edition:v2.1.1
    container_name: plausible_plausible
    hostname: plausible
    restart: always
    volumes:
      - ./config/my-config.exs:/app/my-config.exs:ro
    command: sh -c "sleep 10 && /entrypoint.sh db createdb && /entrypoint.sh db migrate && /entrypoint.sh run"
    depends_on:
      - plausible_db
      - plausible_events_db
    ports:
      - 127.0.0.1:8000:8000
    env_file:
      - plausible-conf.env
```

- `plausible-conf.env` (mail part of it)

```
# MAIL
MAILER_NAME=Plausible Analytics
MAILER_EMAIL=plausible@domain.something
MAILER_ADAPTER=Bamboo.Mua
SMTP_HOST_ADDR=mail.domain.something
SMTP_HOST_PORT=465
SMTP_USER_NAME=plausible@domain.something
SMTP_USER_PWD=***
SMTP_HOST_SSL_ENABLED=false
SMTP_RETRIES=2
EXTRA_CONFIG_PATH=/app/my_config.exs
```

The `my_config.exs` file is in the plausible directory within the subdir `config`.

- `my_config.exs`

```
import Config

config :plausible, Plausible.Mailer,
  ssl: [middlebox_comp_mode: false]
```

This is the error I'm getting:

```
plausible_plausible  | 2024-06-21T21:36:36.346062895Z ERROR! Config provider Config.Reader failed with:
plausible_plausible  | 2024-06-21T21:36:36.348668808Z ** (File.Error) could not read file "/app/my_config.exs": no such file or directory
plausible_plausible  | 2024-06-21T21:36:36.348696048Z     (elixir 1.16.0) lib/file.ex:384: File.read!/1
plausible_plausible  | 2024-06-21T21:36:36.348705823Z     (elixir 1.16.0) lib/config.ex:275: Config.__import__!/1
plausible_plausible  | 2024-06-21T21:36:36.348717744Z     /app/import_extra_config.exs:7: (file)
plausible_plausible  | 2024-06-21T21:36:36.348729248Z     (elixir 1.16.0) src/elixir.erl:405: :elixir.eval_external_handler/3
plausible_plausible  | 2024-06-21T21:36:36.348741347Z     (stdlib 5.2) erl_eval.erl:750: :erl_eval.do_apply/7
plausible_plausible  | 2024-06-21T21:36:36.348752970Z     (stdlib 5.2) erl_eval.erl:136: :erl_eval.exprs/6
plausible_plausible  | 2024-06-21T21:36:36.348762030Z     (elixir 1.16.0) src/elixir.erl:378: :elixir.eval_forms/4
plausible_plausible  | 2024-06-21T21:36:36.348771090Z     (elixir 1.16.0) lib/module/parallel_checker.ex:112: Module.ParallelChecker.verify/1
plausible_plausible  | 2024-06-21T21:36:36.348783249Z
plausible_plausible  | 2024-06-21T21:36:36.349182660Z Runtime terminating during boot ({#{reason=>enoent,path=><<"/app/my_config.exs">>,'__struct__'=>'Elixir.File.Error','__exception__'=>true,action=><<"read file">>},[{'Elixir.File','read!',1,[{file,"lib/file.ex"},{line,384}]},{'Elixir.Config','__import__!',1,[{file,"lib/config.ex"},{line,275}]},{elixir_eval,'__FILE__',1,[{file,"/app/import_extra_config.exs"},{line,7}]},{elixir,eval_external_handler,3,[{file,"src/elixir.erl"},{line,405}]},{erl_eval,do_apply,7,[{file,"erl_eval.erl"},{line,750}]},{erl_eval,exprs,6,[{file,"erl_eval.erl"},{line,136}]},{elixir,eval_forms,4,[{file,"src/elixir.erl"},{line,378}]},{'Elixir.Module.ParallelChecker',verify,1,[{file,"lib/module/parallel_checker.ex"},{line,112}]}]})
plausible_plausible  | 2024-06-21T21:36:36.351497524Z
plausible_plausible exited with code 0
```

**Reply by ruslandoga** [reply]
👋 @wreidlinger 

Seems like a typo issue my-config.exs vs my_config.exs

**Reply by wheeleran** [reply]
Can we please have a solution for those of us who can't or don't know how to make this my_config.exs file? This all used to work before v2, same config, same docker-compose, so wrestling with an abandoned adapter really does feel like a downgrade in experience and functionality.

**Reply by ruslandoga** [reply]
The Erlang version used starting in v2 changed TLS defaults to more secure ones and at the same time became more strict. It's unlikely we would be downgrading to an older Erlang version or making defaults less secure by default for everyone. Mounting a custom config is an easy solution that allows self-hosters to change any configuration just for themselves, I don't see right now how it can be made better.

And we are not wrestling with an abandoned adapter :)

**Reply by wheeleran** [reply]
I was referring to your previous comment in this threat about Bamboo not being maintained...

**Reply by ruslandoga** [reply]
You might be confusing Bamboo (email framework), Bamboo.SMTPAdapter (old adapter), and Bamboo.Mua (new adapter). This thread is about the new adapter. Bamboo is also unmaintained but it is less of an issue since it's not affected by the recent changes in Erlang's TLS.

**Reply by wheeleran** [reply]
Oh, I was think Mua, my mistake

**Reply by wreidlinger** [reply]
> 👋 @wreidlinger
> 
> Seems like a typo issue my-config.exs vs my_config.exs

👋 @ruslandoga you are absolutely right, it was just a stupid typo, guess it was late and my eyes where super tiered!
After correcting the typo the error went ways and I got another error.

```
plausible_plausible  | 2024-06-21T19:23:31.033539969Z 19:23:31.033 [error] Failed to send e-mail
```

But than I switched to the port **587 (STARTTLS)** and everything was running again.
So now my email config in `plausible-conf.env` looks like this:

```bash
# MAIL
MAILER_NAME=Plausible Analytics
MAILER_EMAIL=plausible@domain.something
MAILER_ADAPTER=Bamboo.Mua
SMTP_HOST_ADDR=mail.domain.something
SMTP_HOST_PORT=587
SMTP_USER_NAME=plausible@domain.something
SMTP_USER_PWD=***
SMTP_RETRIES=2
EXTRA_CONFIG_PATH=/app/my_config.exs
```

- `my_config.exs`

```elixir
import Config

config :plausible, Plausible.Mailer,
  ssl: [middlebox_comp_mode: false]
```

You where talking about, that the Erlang version is pretty strict/picky about the TLS/SSL config.
We are using a [mailcow](https://docs.mailcow.email/) email server and I thought the TLS/SSL config is quite correct/good. At least this is what some checking tools told me.

![image](https://github.com/plausible/analytics/assets/100240865/d57d0512-5f2a-4540-af4b-28f3f876a2a5)
Source: https://www.checktls.com/TestReceiver

![image](https://github.com/plausible/analytics/assets/100240865/1c3d6413-1259-46dc-b7af-dcb0522ac585)
Source: https://ssl-tools.net/mailservers

@ruslandoga maybe you have more experience which TLS/SSL config the **Bamboo.Mua** / **Plausible.Mailer** likes or which config should be avoided. Maybe it is worth noticing in the docs section about **Bamboo.Mua**.
For testing the email delivery of the plausible container I was using the proposed syntax from @anttikivi (thanks a lot!).
Maybe it worth asking about a feature request. Would be pretty cool to have a email delivery test function in the plausible web interface... would be nice for admins without to much CLI experience.

Anyways I was testing the email delivery with the following syntax:

```bash
# connect to the plausible container
sudo docker exec -it <name of the container> bin/plausible remote
```

```elixir
# review the current plausible.mailer settings
Application.get_env(:plausible, Plausible.Mailer)
```

```elixir
# create a test email
email = Bamboo.Email.new_email(from: "plausible@yourdomain.something", to: "admin@yourdomain.something", subject: "test", text_body: "test")
```


```elixir
# sends out the test email
Plausible.Mailer.deliver_now(email)
```

@ruslandoga thanks so much for maintaining this great project!!
